### DISCUSSION TITLE
Confirmation when disconnecting from VPN /Do not ask again/

### DISCUSSION THREAD

**Discussion by emilsahakyan** [schwabe/ics-openvpn#1303] [General]
Hi there,

I've automated connecting-disconnecting to my VPN server in certain conditions. For example I use Bixby routines on my Samsung phone to connect to my VPN when I leave home, and disconnect when connected to my home Wi-Fi. I setup this via shortcuts that the APP provides.

It works fine, but when the disconnect is triggered it always asks for confirmation.
I see that it might be do to security reasons (you should know that you're disconnected from the VPN), but I think it would be good to have "Do not ask again" or similar option in the app.
Can this be achieved or implemented ?

Thanks

**Comment by schwabe** [comment]
Yes. It exists: https://github.com/schwabe/ics-openvpn/#controlling-from-external-apps

**Comment by zander-mazda** [comment]
Out of the topic sir schwabe, hope u could help me.

I'm trying to modify a vpn app using apk tool/editor. My source of knowledge about editing is from youtube therefore i am not good at or just a little bit of idea.

My main problem, there are parts that i could not able to edit specific text/string/layout. I dont know where to find that..

Sorry for my poor english. 

Attached is the sample that i wanted to edit.. In an pointed arrow..

Looking forward... Tnx
![AddText_07-08-11 31 21](https://user-images.githubusercontent.com/84930031/124955506-ae833180-e049-11eb-9513-1ea647420d69.jpg)
![AddText_07-08-11 39 47](https://user-images.githubusercontent.com/84930031/124955555-b80c9980-e049-11eb-86ab-d1c2f45ee9dd.jpg)
![AddText_07-08-11 48 13](https://user-images.githubusercontent.com/84930031/124955588-bc38b700-e049-11eb-8a8b-afddd8cfa83c.jpg)

**Comment by CrispedToast** [comment]
It's been a while since this first post, but I have the same issue as the OP - I wish to disconnect using a Bixby Routine on my Samsung phone, but always get the pop-up dialog "Cancel Confirmation - Disconnect the connected VPN/cancel the connection attempt?  Reconnect/Cancel/Disconnect".  I don't understand your answer from March 2021 @schwabe - does it imply that we need to interact via an API to achieve the desired behaviour?

**Reply by schwabe** [reply]
OpenVPN for Android has an API that allows this. I have however no idea what Bixby Routine is doing and will not buy a Samsung phone to figure out how to cater to their non-standard stuff. If someone digs into that and comes up with a reasonable way for that and opens a PR, I am not blocking that but I am also not spending any time and money on supporting/fixing Samsung's broken stuff.

**Reply by CrispedToast** [reply]
Thanks for replying so quickly, and I of course see why you wouldn't want to re-write code for individual requirements or buy handsets from every manufacturer. However just for info, running Open VPN for Android on my Samsung phone works just fine, there is no broken interaction (that I'm aware of) and operating it with fingers/stylus works as intended.

Bixby Routines are really quite good, (like IFTTT for Samsung devices), and add a level of automation which allows interaction with features that apps already offer in their existing UI, namely in this case the 'long press' of the icon.  In the same way that a 'long press' of the Google Chrome icon offers the shortcut options of "New Tab" or "New Incognito Tab", a long press of 'Open VPN for Android' app icon offers the options "Disconnect VPN" and "Connect to [vpn name]" - I presume that is functionality that you have usefully intentionally added.
Bixby Routines can be built to simply simulate the pressing of the "Disconnect VPN" shortcut - this works in so much as it acts identically to how it does when pressing with a finger, (ie as expected). The problem the OP was having, (and me), is that the "Cancel Confirmation" dialogue always pops up after pressing the "Disconnect VPN shortcut" and so the VPN is not actually disconnected unless a user then interacts with the dialog box directly.  Hope that explanation helps.

Thanks anyway for your work on this project - it's a useful and reliable app.

**Reply by schwabe** [reply]
Yes I understand that they might abuse that API. But that API is only intended for Launchers. So I would need to check/see how well it actually behaves and what you can do about it.

**Comment by Kaishley** [comment]
Hello there, just wanted to check on this scenario.
Same like the others, I am using Samsung Routines to automate the connection/disconnection to the VPN.
As a suggestion, maybe there could be a toggle in the Settings to disable the confirmation dialog? Similar to OpenVPN Connect?
Thanks for the good work 😁

**Comment by chiragkrishna** [comment]
I was wondering the same. If there is way to do it. I have to manually select disconnect because of the cancel confirmation thing.

**Comment by untiltheendoftimes** [comment]
It would be great if you could add an option to disable the confirmation upon disconnect

**Comment by pkgf1337** [comment]
I Just recently switched to your App and it's great. Much better than the 'normal' openvpn App. So thanks for that!
I too would Like to have an Option to disable the disconect Dialog. 
Thanks for your Work!!

**Comment by stna1981** [comment]
I have the same problem with my Samsung Routines. Are there any plans to provide such an option to disable the disconnect confirmation dialog?

**Reply by schwabe** [reply]
No. There are intents for automation that serve that purpose.

**Reply by pkgf1337** [reply]
Fine.
If anyone else has this Problem, here is how I solved it for me.
instead of disconnecting (ie when I'm in my home wifi) my automation just connects to a second connection which does not route any traffic (like a dummy vpn). switching from one connection to another doesn't require a confirmation, this way I avoid the disconnect dialog and don't need to every touch the app regardless of where I am. should work with all automation apps.

**Reply by stna1981** [reply]
That sounds interesting, do you have some advice or is there a guide how to create such a dummy VPN? I created my regular one with PiVPN so I might be lacking some deeper knowledge about how to create such a VPN...

**Reply by chiragkrishna** [reply]
https://github.com/zaneschepke/wgtunnel

Thank me later.

**Reply by schwabe** [reply]
@chiragkrishna that is not helpful in the least to say that people should use a random wireguard app instead.

**Reply by chiragkrishna** [reply]
No disrespect, but i waited a long time to disable that annoying confirmation dialogue. 
https://github.com/schwabe/ics-openvpn/issues/1695

That app finally solves the problem with inbuilt autotunneling feature.

**Reply by chuzmunguz** [reply]
@schwabe the app is amazing and does great good to everyone that uses it, but I have to admit that I also have been waiting to get a way to disable that notification for the same reason. It just breaks the automation and is very annoying.

**Reply by schwabe** [reply]
Unless someone who actually has a Samsung can figure out what Samsung Automations actually does and how an app can properly support it, there is nothing from my side that I can do. I will not buy a Samsung phone just to see what non-standard stuff they are doing. This is still a spare time project. If someone else wants to do the effort and do the work and provide a proper PR with an explaination how this interacts with Samsung, I will include it but in the mean time, I cannot implement something that I cannot test.

**Reply by chuzmunguz** [reply]
I understand your concerns, but you don't need a Samsung phone to test what we're asking to be implemented.

All that is needed is a setting that allow us to disable the "Cancel notification" message after you tapped on the profile to disable the vpn connection.

I'm sure that, as the software dev, you know what I'm referring to.

That's the same behaviour we get when automate the vpn disconnection, just like we have tapped on the vpn profile to disable the connection. The problem is that that notification prevents the disconnection to be fully automated, because human confirmation is needed to actually shut down the vpn.

**Reply by chiragkrishna** [reply]
@schwabe we are not asking much man, just add a small tick box in **Cancel Confirmation**
```
[x] Dont Ask Again
```
the screenshots are here https://github.com/schwabe/ics-openvpn/issues/1695

**Reply by schwabe** [reply]
Unless someone points me to a guide where Samsung describes for developers their automation like a document that describes how make apps work well with that, I am not going to change anything. There are multiple ways to trigger a VPN disconnect (at least 5) and it is completely unclear which one of these Samsung's stuff is using. What is clear is that they are NOT using one of the two  ways (vai AIDL/via Acitivities) that the app already implements to allow automation via other apps.

**Reply by chiragkrishna** [reply]
According to [this](https://forum.developer.samsung.com/t/routines-and-modes-custom-task-support/33868)
Samsung Automation uses [app shortcuts](https://developer.android.com/develop/ui/views/launch/shortcuts/creating-shortcuts)

The automation on Samsung works with your app, only problem is the manual **Cancel Confirmation** popup

**Reply by chuzmunguz** [reply]
Well, I'm not sure what more do you need... everybody already described what the issue is, and what is needed to solve it, and as a sw dev myself, cannot understand what the difficulty is to implement a dialog bypass.

It is in your right if you don't want to implement the change, but please, just admit it, don't keep making it like it is something overcomplicated when it just requires an additional setting to turn off the "Cancel notification" when you try to disconnect the vpn.

I'm sure that you know which of that 5 ways to turn of the vpn connection use the dialog, and bypass it.

Sorry for the rant but I don't understant why are you offering so much resistance to adapt the app to something that is new and used more every day. It is not reasonable to expect that a major phone brand adapts it's OS automation process to every single app that exists, unfortunately it has to be the other way around.

**Reply by schwabe** [reply]
Samsung ROMs have given me more than enough trouble over the years. Using app shorts and the disconnect via app shortcuts is a completely different code path from using disconnect via the notification. These two code paths are even starting in different processes of the app. 

Using app shortcuts for automation is a clear abuse of that feature and apparently because it is Samsung and that is a big company that is okay. That they are not able to give developers a way to actually expose actions to their shortcut system in a sensible way but use this hacky way does not inspire confidence.

And this is my spare time project. I have no interest currently investing time and money to figure out if that can be solved in a proper way like if Samsung calls the app in a way with the shortcuts that the app can determine if it is an automation call vs a user call or if Samsung routes has better support. So this doesn't get done. Plain and simple. I really don't like the way that apparently I need to make my app shorts designed for their normal intended use but also for the weird Samsung routines use case that is not even properly document. And sorry, a single random post in a forum does not really count as a documentation.

I am basically pressured by user like you to change my app to support a feature that I cannot test and that is also not documented. Yes, it sounds like a simple change but then there is the next thing that breaks with Samsung routines and then next thing and then I next thing. That is not something I want to deal with in my spare time. And my experience with Samsung software is really bad. That is why I am saying that someone should look into that if they really want this being fixed.

**Reply by chuzmunguz** [reply]
Ok, fair enough. Thank you for your dedication and contribution to the FOSS community.

**Reply by chiragkrishna** [reply]
I think there is a miss communication here. 
**This feature has nothing to do with a phone's brand**
Let it be Samsung, Motorola, Nokia, Google etc etc.. thay have nothing to do here,
The request is simple, let the user have an option to disable the confirmation. Any ways thanks for the efforts. I've moved on to wireguard anyways.

**Reply by schwabe** [reply]
@chiragkrishna I like to keep the confirmation. I don't think the extra click is hurting anyone. The extra click only hurts Samsung Automation. You might not agree with my decision here but that is fine.

**Reply by 662KDV** [reply]
@schwabe Since there's no option to automatically disconnect from VPN while connected to home wifi, I'm also using samsung's modes & routines to accomplish this goal. Is any of these two ever going to be implemented?
Unfortunately removing notification permission also does not fix the issue.

Otherwise congrats, this is a solid app, it does it's job. Reconnection is fast when wifi/cell reception is spotty. (compared to android native IPsec client)

**Reply by schwabe** [reply]
The beta version has a checkbox that allows you to diable all confirmations

**Reply by pkgf1337** [reply]
Works perfectly now, thanks @schwabe

**Reply by emilsahakyan** [reply]
I switched to beta version - works great, thank you @schwabe
